src/dbnode/discovery/config.go

```diff
@@ -0,0 +1,234 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package discovery provides discovery configuration.
+package discovery
+
+import (
+	"errors"
+	"fmt"
+
+	etcdclient "github.com/m3db/m3/src/cluster/client/etcd"
+	"github.com/m3db/m3/src/dbnode/environment"
+)
+
+const (
+	defaultEnvironment                   = "default_env"
+	defaultZone                          = "embedded"
+	defaultM3DBService                   = "m3db"
+	defaultM3AggregatorService           = "m3aggregator"
+	defaultCacheDirectory                = "/var/lib/m3kv"
+	defaultSingleNodeClusterEndpoint     = "127.0.0.1:2379"
+	defaultSingleNodeClusterSeedEndpoint = "http://127.0.0.1:2380"
+)
+
+var validDiscoveryConfigTypes = []ConfigurationType{
+	ConfigType,
+	M3DBSingleNodeType,
+	M3DBClusterType,
+	M3AggregatorClusterType,
+}
+
+// ConfigurationType defines the type of discovery configuration.
+type ConfigurationType uint
+
+const (
+	// ConfigType defines a generic definition for service discovery via etcd.
+	ConfigType ConfigurationType = iota
+	// M3DBSingleNodeType defines configuration for a single M3DB node via etcd.
+	M3DBSingleNodeType
+	// M3DBClusterType defines M3DB discovery via etcd.
+	M3DBClusterType
+	// M3AggregatorClusterType defines M3DB discovery via etcd.
+	M3AggregatorClusterType
+)
+
+// UnmarshalYAML unmarshals an ConfigurationType into a valid type from string.
+func (t *ConfigurationType) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var str string
+	if err := unmarshal(&str); err != nil {
+		return err
+	}
+
+	// If unspecified, use default mode.
+	if str == "" {
+		*t = ConfigType
+
+		return nil
+	}
+
+	for _, valid := range validDiscoveryConfigTypes {
+		if str == valid.String() {
+			*t = valid
+
+			return nil
+		}
+	}
+
+	return fmt.Errorf("invalid ConfigurationType '%s' valid types are: %s",
+		str, validDiscoveryConfigTypes)
+}
+
+// String returns the discovery configuration type as a string.
+func (t ConfigurationType) String() string {
+	switch t {
+	case ConfigType:
+		return "config"
+	case M3DBSingleNodeType:
+		return "m3db_single_node"
+	case M3DBClusterType:
+		return "m3db_cluster"
```

Comment:
Do we need to disambiguate between these two cases at this level? Can we do the right thing based on whether the provided cluster config has 1 or >1 entries?

Reply:
Unfortunately not, for m3db_single_node case the etcd configuration is not specified (to make this reduced config possible). When m3db_single_node is set it knows without specifying the local host address as the etcd to connect to (i.e. you haven't connected to etcd yet to infer that there's a single node present, you need to inform DB that you are expecting a single node and then it can follow the behavior of connecting to etcd on the embedded DB node instead of making you type out in config the etcd address).

Reply:
Ah, I see what you mean now - makes sense.